lightning/src/ln/outbound_payment.rs

```diff
@@ -504,10 +495,12 @@ impl OutboundPayments {
 		loop {
 			let mut outbounds = self.pending_outbound_payments.lock().unwrap();
 			let mut retry_id_route_params = None;
-			for (pmt_id, pmt) in outbounds.iter_mut() {
-				if pmt.is_auto_retryable_now() {
-					if let PendingOutboundPayment::Retryable { pending_amt_msat, total_msat, payment_params: Some(params), .. } = pmt {
-						if pending_amt_msat < total_msat {
+			outbounds.retain(|pmt_id, pmt| {
+				let mut retain = true;
+				let auto_retryable = pmt.is_auto_retryable_now();
+				if let PendingOutboundPayment::Retryable { pending_amt_msat, total_msat, payment_params: Some(params), .. } = pmt {
```

Comment:
I'm confused what's going on in this loop (`outbounds.retain(...)`). For each outbound payment, it looks to be checking if it's valid to retry, and either marking it abandoned and pushing a `PaymentFailed` event or setting `retry_id_route_params` to retry the payment. I'm mainly confused because it looks like if there's multiple retryable payments, this will just set `retry_id_route_params` multiple times and afterwards only retry whatever single payment it was last set to. I'd imagine it'd make more sense want to retry each retryable payment that it finds from this iteration (instead of looping through all payments to just retry a single one) right? Not sure if I'm understanding this correctly

Reply:
This was something I meant to revisit but forgot, thanks. I'm still not 100% sure of the new solution since it adds an extra lock of `pending_outbound_payments`, but it seems overall better to me for the reasons you specified. 

We can't retry each payment as we go because we need to drop the `pending_outbound_payments` lock each time we call `pay_internal`. It is a bit weird though. See this discussion for more info: https://github.com/lightningdevkit/rust-lightning/pull/1916#discussion_r1085862125

Reply:
> We can't retry each payment as we go because we need to drop the pending_outbound_payments lock each time we call pay_internal. It is a bit weird though. See this discussion for more info: https://github.com/lightningdevkit/rust-lightning/pull/1916#discussion_r1085862125

Ah gotcha.

I think the new solution is an improvement at least in readability with the iteration for retrying/lock dropping and abandoning of payments being separate, but yea, this seems tough heh :)